sui/contract/sources/constant_product.move

```diff
@@ -7,78 +7,133 @@
+    const EPoolAlreadyExists: u64 = 2;
+    const EPoolEmpty: u64 = 3;
+    const EPoolOverflow: u64 = 4;
+    const EMinSwap: u64 = 5;
 
     const MAX_BPS: u128 = 10000;
 
-    struct LP<phantom X, phantom Y, phantom W> has drop {}
+    /***********/
+    /* Objects */
+    /***********/
+
+    struct Pools has key {
+        id: UID,
+        pools: VecSet<TypeName>
+    }
 
-    struct Pool<phantom X, phantom Y, phantom W> has key {
+    struct Pool<phantom X, phantom Y> has key {
         id: UID,
         reserve_x: Coin<X>,
         reserve_y: Coin<Y>,
-        supply_lp: Supply<LP<X, Y, W>>,
-        fee_bps: u64,
+        supply_lp: Supply<LP<X, Y>>,
+        fee_bps: u64
+    }
+
+    struct LP<phantom X, phantom Y> has drop {}
+
+    /**********/
+    /* Events */
+    /**********/
+
+    struct PoolsCreated has copy, drop {
+        id: ID
     }
 
-    public fun reserve_x<X, Y, W>(self: &Pool<X, Y, W>): &Coin<X> {
+    struct PoolCreated<phantom X, phantom Y> has copy, drop {
+        id: ID
+    }
+
+    struct Swapped<phantom IN, phantom OUT> has copy, drop {
+        amount_in: u64,
+        amount_out: u64 
+    }
+
+    struct LiquidityAdded<phantom X, phantom Y> has copy, drop {
+        amount_added_x: u64,
+        amount_added_y: u64,
+        amount_minted_lp: u64
+    }
+
+    struct LiquidityRemoved<phantom X, phantom Y> has copy, drop {
+        amount_removed_x: u64,
+        amount_removed_y: u64,
+        amount_burned_lp: u64
+    }
+
+    /***************/
+    /* Module init */
+    /***************/
+
+    /// `Pools` can only be created once on deploy
+    fun init(ctx: &mut TxContext) {
+        let id = object::new(ctx);
+        let inner_id = object::uid_to_inner(&id);
+        let pools = Pools {
+            id,
+            pools: vec_set::empty()
+        };
+        event::emit(PoolsCreated { id: inner_id });
+        transfer::share_object(pools);
+    }
+
+    /***********************/
+    /* Getters and setters */
+    /***********************/
+
+    public fun reserve_x<X, Y>(self: &Pool<X, Y>): &Coin<X> {
         &self.reserve_x
     }
 
-    public fun reserve_y<X, Y, W>(self: &Pool<X, Y, W>): &Coin<Y> {
+    public fun reserve_y<X, Y>(self: &Pool<X, Y>): &Coin<Y> {
         &self.reserve_y
     }
 
-    public fun supply_lp<X, Y, W>(self: &Pool<X, Y, W>): &Supply<LP<X, Y, W>> {
+    public fun supply_lp<X, Y>(self: &Pool<X, Y>): &Supply<LP<X, Y>> {
         &self.supply_lp
     }
 
-    public fun fee_bps<X, Y, W>(self: &Pool<X, Y, W>): u64 {
+    public fun fee_bps<X, Y>(self: &Pool<X, Y>): u64 {
         self.fee_bps
     }
 
-    entry fun create_pool_<X, Y, W: drop>(
-        _: W,
-        fee_bps: u64,
-        ctx: &mut TxContext
-    ) {
-        create_pool<X, Y, W>(_,fee_bps, ctx);
-    }
+    /*******************/
+    /* Entry functions */
+    /*******************/
 
-    public fun create_pool<X, Y, W: drop>(
-        _: W,
+    entry fun create_pool_<X, Y>(
```

Comment:
What's the reason to omit `public`?

Reply:
`public` used to be implied for `entry fun` but just saw they changed the visibility rules recently to be able to call it from another contract